### PR TITLE
test-dump-counters: Test dump-counters via unix-command socket

### PR DESCRIPTION
**Pull request by awelzel**
## Ticket

Test for: https://github.com/OISF/suricata/pull/10468

Redmine ticket: https://redmine.openinfosecfoundation.org/issues/6732

### CODE REVIEW

**Comment by jasonish** [comment]
Interesting approach, but I do have a few concerns..
- First, it fails in GitHub CI and when I run locally.
- Second, for 8.0 I hope to have the Python Suricata SC gone. Could this be written in terms of the `suricatasc` command? I plan to have a CLI to replace that, but PR work has only been drafts as of now.

I think simile semi-live tests have been desired in the past, so need to think about how this fits into the bigger picture I think. But getting it passing would be a good start.

```
Executing: timeout -k 1 10 /home/jason/oisf/dev/suricata/review/src/suricata -v -l /home/jason/oisf/dev/suricata-verify/review/tests/test-dump-counters/output -c /home/jason/oisf/dev/suricata/review/suricata.yaml --pcap=lo --runmode=workers --pidfile=/home/jason/oisf/dev/suricata-verify/review/tests/test-dump-counters/output/suricata.pid --set capture.disable-offloading=false --set capture.checksum-validation=none --set pcap.1.threads=2 --set flow.managers=3 --set flow.recyclers=5 --set stats.interval=1 --set unix-command.filename=/home/jason/oisf/dev/suricata-verify/review/tests/test-dump-counters/output/suricata.sock &
PYTHONPATH=/home/jason/oisf/dev/suricata/review/python /home/jason/oisf/dev/suricata-verify/review/tests/test-dump-counters/test-dump-counters.py --pidfile /home/jason/oisf/dev/suricata-verify/review/tests/test-dump-counters/output/suricata.pid --suricata-socket /home/jason/oisf/dev/suricata-verify/review/tests/test-dump-counters/output/suricata.sock > /home/jason/oisf/dev/suricata-verify/review/tests/test-dump-counters/output/dump-counters.json
Notice: suricata: This is Suricata version 8.0.0-dev running in SYSTEM mode [LogVersion:suricata.c:1147]
Info: cpu: CPUs/cores online: 20 [UtilCpuPrintSummary:util-cpu.c:182]
Info: suricata: Setting engine mode to IDS mode by default [PostConfLoadedSetup:suricata.c:2684]
Info: exception-policy: master exception-policy set to: auto [ExceptionPolicyMasterParse:util-exception-policy.c:200]
Info: ioctl: lo: MTU 65536 [GetIfaceMTU:util-ioctl.c:100]
Info: suricata: Preparing unexpected signal handling [InitSignalHandler:suricata.c:2186]
Info: conf: Running in live mode, activating unix socket [ConfUnixSocketIsEnable:util-conf.c:154]
Info: logopenfile: fast output device (regular) initialized: fast.log [SCConfLogOpenGeneric:util-logopenfile.c:616]
Info: logopenfile: eve-log output device (regular) initialized: eve.json [SCConfLogOpenGeneric:util-logopenfile.c:616]
Info: logopenfile: stats output device (regular) initialized: stats.log [SCConfLogOpenGeneric:util-logopenfile.c:616]
Info: classification-config: Added "43" classification types from the classification file [SCClassConfParseFile:util-classification-config.c:370]
Info: reference-config: Added "19" reference types from the reference.config file [SCRConfParseFile:util-reference-config.c:342]
INFO:root:pidfile and socket appeared
INFO:root:Failed to connect: L178: [Errno 2] No such file or directory
INFO:root:Failed to connect: L178: [Errno 2] No such file or directory
INFO:root:Failed to connect: L178: [Errno 2] No such file or directory
INFO:root:Failed to connect: L178: [Errno 2] No such file or directory
INFO:root:Failed to connect: L178: [Errno 2] No such file or directory
INFO:root:Failed to connect: L178: [Errno 2] No such file or directory
INFO:root:Failed to connect: L178: [Errno 2] No such file or directory
...
INFO:root:Stopping Suricata
===> test-dump-counters: FAILED: got exit code 2, expected 0

PASSED:  0
FAILED:  1
SKIPPED: 0
```

**Comment by awelzel** [comment]
> Thanks, interesting approach, but I do have a few concerns..

:sweat_smile: 

> First, it fails in GitHub CI and when I run locally.

Did you pick up the fix from https://github.com/OISF/suricata/pull/10468 ? I broke stats in 7.0.x and master, so the CI failures are valid I think. Do I just put the "requires suricata pr" label on it?

> Second, for 8.0 I hope to have the Python Suricata SC gone. Could this be written in terms of the suricatasc command?

Oh. Hmm. I guess that could be done. I opted for a bit of Python convenience compared to the shell scripting involved, but I can give that a shot.

> so need to think about how this fits into the bigger picture I think. 

Yes, totally agree. I was thinking of extending the yaml with some keywords to allow running `command` in the background and executing commands, but I think that's a project by itself. Seems useful though even if only testing interactions via the unix socket though (rule reloading, pcap submission, ...).

**Comment by jasonish** [comment]
Ok, so figured out the issues.

Running on my workstation, and if I happen to be running from my Suricata build directory, which has also been installed, and ASAN enabled, this times out before all the rules are loaded and Suricata is listening.  Providing an empty rule could prevent this timeout.

I guess my `lo` is sufficiently busy that I get a `wrong_thread` errors in the JSON, and these are not yet in the schema, so the test fails out there. Kind of a separate issue, but still needs a fix.

**Comment by jasonish** [comment]
This will fix the schema issue: https://github.com/OISF/suricata/pull/10534

File: tests/test-dump-counters/test-dump-counters.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        if result["return"] != "OK":
+            logging.info("Non-OK result: %s", result)
+            time.sleep(poll_interval)
+            continue
+
+        print(json.dumps(result))
+        exit_code = 0
+        break
+
+    logging.info("Stopping Suricata")
```
Comment (victorjulien) [inline tests/test-dump-counters/test-dump-counters.py:63]: can just send the shutdown command over the socket here?

File: tests/test-dump-counters/test-dump-counters.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    args = parser.parse_args()
+
+    logging.basicConfig(level=logging.DEBUG if args.verbose else logging.INFO)
+
+    poll_interval = args.poll_interval
+    end_time = time.time() + args.timeout
+    suricata_pid = None
+    exit_code = 1
+
+    while time.time() < end_time:
```
Comment (victorjulien) [inline tests/test-dump-counters/test-dump-counters.py:30]: in a private bash script unix test I loop until the `version` unix command succeeds

**Comment by awelzel** [comment]
Replaced by https://github.com/OISF/suricata-verify/pull/1676

**Pull request closed without merging**